### PR TITLE
Skeleton of drafter

### PR DESCRIPTION
**Pull request by klokane**
__Please still don't push into master__

It is preview, there are still missing some features like integration tests, connect to CI, connect to Hipchat, etc

Not tested to compile on windows && mac.

Please review @zdne and @pksunkara 

Comments are welcome!!

### CODE REVIEW

File: src/drafter.cc [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+struct fstream_selector<std::istream>{
+  typedef std::ifstream stream_type;
+};
+
+template<> 
+struct fstream_selector<std::ostream>{
+  typedef std::ofstream stream_type;
+};
+
+
```
Comment (pksunkara) [inline src/drafter.cc]: Extra lines.

File: src/drafter.cc [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+      std::cerr << "fatal: unable to open file '" << file << "'\n";
+      exit(EXIT_FAILURE);
+    }
+
+    return stream;
+}
+
+void Serialize(const std::string& out, 
+    const sos::Object& object, 
+    sos::Serialize* serializer)
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: Alignment on the above two lines.

File: src/drafter.cc [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    std::stringstream ss;
+
+    ss << "<input file>\n\n";
+    ss << "API Blueprint Parser\n";
+    ss << "If called without <input file>, 'drafter' will listen on stdin.\n";
+
+    parser.footer(ss.str());
+}
+
+#define DRAFTER_VERSION_STRING "x.x.x"
```
Comment (pksunkara) [inline src/drafter.cc]: Shouldn't this be at the beginning of the file after `#includes`? Just curious if there is any reason for it be in the middle here.

Comment (klokane) [inline src/drafter.cc]: it will be replace by "version.h" like snowcrash
Now it is just fake string

**Comment by klokane** [comment]
Now it should be ready to merge, full implementation of original commandline snowcrash functionality is complete.

So please @zdne @pksunkara for review

File: configure [hunk 4]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
```
Comment (pksunkara) [inline configure]: I think this should be `python2`. See apiaryio/snowcrash#277

File: configure [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+parser.add_option("--dest-cpu",
+    action="store",
+    dest="dest_cpu",
+    help="CPU architecture to build for. Valid values are: ia32, x64.")
+
+#parser.add_option("--shared",
+#    action="store_true",
+#    dest="shared",
+#    help="Build and use shared libsnowcrash instead of static one.")
```
Comment (pksunkara) [inline configure]: I don't know the purpose of this commented out part. Can we remove it completely if it is not going to be used? /cc @zdne

Comment (klokane) [inline configure]: It is backport form snowcrash. We will need it later probably for library interface

File: features/support/env.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+require 'aruba/cucumber'
+
+Before do
+  @dirs << "../../features/fixtures"
+  
+  ENV['PATH'] = "./bin#{File::PATH_SEPARATOR}#{ENV['PATH']}"  
+end
```
Comment (pksunkara) [inline features/support/env.rb]: nitpick: no newline at of

Comment (klokane) [inline features/support/env.rb]: 1:1 copy from snowcrash :) will fix

File: src/Version.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+#ifndef DRAFTER_VERSION_H
+#define DRAFTER_VERSION_H
+
+#define DRAFTER_MAJOR_VERSION 0
+#define DRAFTER_MINOR_VERSION 1
+#define DRAFTER_PATCH_VERSION 0
+
+#define DRAFTER_VERSION_IS_RELEASE 0
+
+#ifndef DRAFTER_STRINGIFY
```
Comment (pksunkara) [inline src/Version.h]: hmm.. I am not sure, but we don't need this `ifndef`. Right? `ifndef DRAFTER_VERSION_H` would take care of this.

Comment (klokane) [inline src/Version.h]: It's too copied from snowcrash just s/SNOWCRASH/DRAFTER/

But I agree it can be removed

File: src/config.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
+#ifndef CONFIG_H_PIBRI8ND
+#define CONFIG_H_PIBRI8ND
```
Comment (pksunkara) [inline src/config.h]: I am really curious about where this came from. Even though I have no issue, for consistency sake, we should probably use `DRAFTER_CONFIG_H`.

File: src/config.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+//
+//  Created by Jiri Kratochvil on 2015-02-13
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+//
+//
+
+#include <string>
+
+
```
Comment (pksunkara) [inline src/config.h]: Extra empty line.

File: src/config.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+namespace config {
```
Comment (pksunkara) [inline src/config.h]: Isn't `namespace drafter` a better name?

File: src/drafter.cc [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    return_type operator()() const
+    { 
+        return return_type(std_io<T>()(), dummy_deleter<stream_type>()); 
+    }
+};
+
+/**
+ *  \brief functor select fstream based on iostream
+ */
+template <typename Stream> struct to_fstream;
```
Comment (pksunkara) [inline src/drafter.cc]: `<typename T>`. Excuse my ignorance if you used `Stream` intentionally.

File: src/drafter.cc [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+ *  In similar way it work for std::ostream
+ *
+ *  \param template<T> type of returned stream
+ *  \param file - name of file to open for read/write if empty use standart input/output
+ *
+ */
+template<typename T>
+std::tr1::shared_ptr<T> CreateStreamFromName(const std::string& file)
+{
+    if(file.empty()) {
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: space between `if` and `(`

**Comment by pksunkara** [comment]
Sorry about the *nitpicks*, but I find that they increase the readability of the code.

File: src/drafter.cc [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+/**
+ *  \brief  return instance sos::Serializer based on \param `format`
+ *
+ *  \param format - output format for serialization
+ */
+
+sos::Serialize* CreateSerializer(const std::string& format)
+{
+    if(format == "json") {
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: space between `if` and `(`

File: src/drafter.cc [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        return new sos::SerializeJSON;
+    } else if(format == "yaml") {
```
Comment (pksunkara) [inline src/drafter.cc]: ditto.

File: src/drafter.cc (hunk 11 above)
Comment (klokane) [inline src/drafter.cc]: Yes, it is intentionally, to hint paremeter must be iostream, but you are right it differs from rest of file whewre is used <T>. 
Will fix

File: src/drafter.cc [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    std::cerr << "fatal: unknow serialization format: '" << format << "'\n";
+    exit(EXIT_FAILURE);
+}
+
+/**
+ * \brief Serialize sos::Object into stream
+ */
+void Serialization(std::tr1::shared_ptr<std::ostream> stream,
+        const sos::Object& object,
+        sos::Serialize* serializer)
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: Alignment of the above two lines.

File: src/drafter.cc [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+
+int main(int argc, const char *argv[])
+{
+    Config config; 
+    ParseCommadLineOptions(argc, argv, config);
+
+    sc::BlueprintParserOptions options = 0;  // Or snowcrash::RequireBlueprintNameOption
+    if(!config.sourceMap.empty()) {
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: space between `if` and `(`

File: src/drafter.cc [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+
+    sc::ParseResult<sc::Blueprint> blueprint;
+    sc::parse(inputStream.str(), options, blueprint);
+
+    if (!config.validate) {  // not just validate -> we will serialize
+        sos::Serialize* serializer = CreateSerializer(config.format);
+
+        Serialization(CreateStreamFromName<std::ostream>(config.output),
+            snowcrash::WrapBlueprint(blueprint.node), 
+            serializer);
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: Alignment on the above two lines.

File: src/drafter.cc [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+        Serialization(CreateStreamFromName<std::ostream>(config.sourceMap),
+            snowcrash::WrapBlueprintSourcemap(blueprint.sourceMap),
+            serializer);
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: Alignment on the above two lines.

File: src/reporting.cc [hunk 19]
```diff
@@ -0,0 +1,164 @@
+#include "reporting.h"
+
+#include <algorithm>
+#include <iostream>
```
Comment (pksunkara) [inline src/reporting.cc]: hmm.. why are these `includes` before the comment header? Other files have them after the comment header.

File: src/reporting.h [hunk 20]
```diff
@@ -0,0 +1,30 @@
+#ifndef REPORTING_H_JHA2MUHM
+#define REPORTING_H_JHA2MUHM
```
Comment (pksunkara) [inline src/reporting.h]: Similarly, `DRAFTER_REPORTING_H`

File: src/drafter.cc [hunk 21]
```diff
@@ -0,0 +1,204 @@
+// vi:cin:et:sw=4 ts=4
+//
+//  drafter.cc
+//
+//  Created by Jiri Kratochvil on 2015-02-11
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+//
+//
```
Comment (pksunkara) [inline src/drafter.cc]: nitpick: These 2 extra empty comment lines are not needed. Just 1 empty comment line after the `Copyright` line is enough. Similarly in all the other files.

**Comment by pksunkara** [comment]
@klokane Reviewed. Just minor things. Everything else looking good. Great job! :smile:

File: src/reporting.cc (hunk 19 above)
Comment (klokane) [inline src/reporting.cc]: accident

File: src/config.h (hunk 8 above)
Comment (klokane) [inline src/config.h]: _PIBRI8ND is random string to be "unique" but i will rename it in way you suggest

File: src/config.h (hunk 10 above)
Comment (klokane) [inline src/config.h]: I will move it into config.cc need no to be public in headers

**Comment by klokane** [comment]
Fixed

File: src/drafter.cc (hunk 15 above)
Comment (pksunkara) [inline src/drafter.cc]:
```cpp
void Serialization(std::tr1::shared_ptr<std::ostream> stream,
                   const sos::Object& object,
                   sos::Serialize* serializer)
```

**Comment by pksunkara** [comment]
Please also fix the 3-4 alignment issues left and I can merge this.

**Comment by klokane** [comment]
Fixed, pushed with force.

But please explain me, what do you mean about 
features/support/env.rb
"no newline at eof" 

It is missing? rt it should be removed?

File: src/drafter.cc (hunk 15 above)
Comment (pksunkara) [inline src/drafter.cc]: @klokane, Only this is left for now.

**Comment by klokane** [comment]
complete, hope...

**Comment by pksunkara** [comment]
Thanks :)

**Comment by zdne** [comment]
@pksunkara @klokane is this already tested in CircleCI? If so please add badge to the README. Thanks

**Comment by pksunkara** [comment]
@zdne No, no circleCI tests yet.

File: configure (hunk 4 above)
Comment (zdne) [inline configure]: > I think this should be python2. See apiaryio/snowcrash#277

As you have correctly assumed. Hardcoding python2 is not the correct solution. Does not work on systems with Python 2.7.6 only.

**Pull request merged**